### DISCUSSION TITLE
Constantly read parameters from an external file to manipulate functions in Stellarium using its script engine.

### DISCUSSION THREAD

**Discussion by FogoVoar** [Stellarium/stellarium#3760] [Q&A] [answered]
I have a project in mind that involves manipulating the yaw and pitch of the Stellarium camera using values ​​from another program (which can be saved in a file using any format).

I realized that for security reasons Stellarium's script engine doesn't allow manipulating files, but it does allow including them, so this was my attempt:

With my external program I create the following `input.inc` where the values ​​change every second:
```
var pitch = 45;
var yaw = 90;
```

In the Stellarium console I try to repeatedly include the file:
```
function refresh() {
	include("input.inc");
	core.moveToAltAzi(pitch, yaw, 0);
	core.wait(5);
	refresh();
}
refresh()
```
all I get is:
```
script error: 'Reference'  @ line 2: ReferenceError: include is not defined
```
which is weird because include works (once) if placed outside the function scope.
Does anyone have any idea of ​​a better approach?

**Comment by gzotti** [comment]
Obvious solution: place include outside the function scope. Maybe rewrite as `refresh(pitch, yaw);`

**Reply by FogoVoar** [reply]
```
include("input.inc");
function refresh() {
	core.moveToAltAzi(pitch, yaw, 0);
	core.wait(2);
	refresh();
}
refresh()
```
in this case include will only be included once and successive changes to input.inc will have no effect.
Rewriting it as `refresh(pitch, yaw);` would put it in a locked loop where the pitch and yaw value would never change.

**Reply by gzotti** [reply]
I did not see the application context of the whole thing. include is a non-standard preprocessing command executed before actually starting the script, so it cannot be used to read dynamically changing files.

**Comment by 10110111** [comment]
You could just generate the whole function body instead of just the declarations of variables.

**Reply by FogoVoar** [reply]
I tried this at first but I didn't find a way to run the script constantly...

**Reply by 10110111** [reply]
Oh, actually indeed, `include()` will happen once per script invocation. I wonder though, how did you come to the conclusion that there's a restriction of reading files? I don't remember having seen it being put into Stellarium. There is a restriction on writing files with absolute paths (which is controlled by a flag), but there doesn't seem to be any for reading.

**Reply by FogoVoar** [reply]
I got the information from [here](https://github.com/Stellarium/stellarium/discussions/3243#discussioncomment-6005206)
but if there is a way to read files it would make everything easier! :)

**Reply by gzotti** [reply]
OK, if there is a general way of reading just any input files from anywhere, I am unaware of it. This example may be much simpler by just calling, from the external program, the RemoteControl API with curl (or of course directly).

**Reply by 10110111** [reply]
Apparently, the typical way to read files from JS is by using `XMLHttpRequest`. But it's not available in QtScript. So it's impossible not for security reasons, but simply due to a lack of implementation.

**Reply by FogoVoar** [reply]
> OK, if there is a general way of reading just any input files from anywhere, I am unaware of it. This example may be much simpler by just calling, from the external program, the RemoteControl API with curl (or of course directly).

I wasn't aware of the RemoteControl API, I enabled the plugin and started the server on port 20180, with the curl command it seems to be able to access the API, but I have not yet had success in manipulating stellarium or obtaining data.

this is the command I'm trying:
`curl --data 'j2000=[0.5,0.3,0.2]' http://192.168.0.104:20180/api/main/view --output log.txt`
it returns `curl: (1) Received HTTP/0.9 when not allowed`

by appending `--http0.9` like:
`curl --http0.9 --data 'j2000=[0.5,0.3,0.2]' http://192.168.0.104:20180/api/main/view --output log.txt`
the log.txt returns:
```
 5StellariumSyncPluginProtocol     È°ª
#Gm˜êå»•p‡l     invalid message type 80
```
(it also only appears to work if I specify my machine address directly, not with localhost).

**Reply by gzotti** [reply]
I don't understand why there is a complaint about  HTTP/0.9, never seen that. I can call 
`curl --data 'j2000=[0.5,0.3,0.2]' http://192.168.xxx.xxx:8090/api/main/view`
from my WSL console (Ubuntu 22.04.04, curl 7.81.0) and it just works. --http0.9 or --http1.0 also work, --http2.0 does not work.

**Reply by FogoVoar** [reply]
Hm, weird, I'm running WSL with Ubuntu 22.04.3 and curl 7.81.0 (and Stellarium 24.1).
Thanks anyway! for now I'm going to shelve the idea, hopefully someone in the future can revive this thread with some workaround.

**Comment by FogoVoar** [comment, marked as answer]
As @gzotti suggested, it is possible to achieve this by using the [RemoteControl](https://stellarium.org/doc/head/remoteControlApi.html) plugin to start a server with an API for Stellarium functions.
Press F2 > Plugins > Remote Control > Load at startup > configure > [x] Server enabled
If the configure button is unavailable, simply restart Stellarium.

To send the commands you will need to send HTTP requests using something like [fetch](https://nodejs.org/dist/latest-v18.x/docs/api/globals.html) or [axios](https://www.npmjs.com/package/axios) (both comes by default in nodejs), or [curl](https://[curl](https://curl.se/).se/).

In the RemoteControl documentation you can find the "view" field which has 3 functions to rotate the camera:
```
    "j2000" : "[x,y,z]",
    "jNow"  : "[x,y,z]",
    "altAz" : "[x,y,z]"
```
To call the function with parameters you must follow this format:
`j2000=[0.5,0.3,0.2]`

Example with curl:
`curl --data 'j2000=[0.5,0.3,0.2]' http://192.168.xxx.xxx:8090/api/main/view` In the IP field you will define the IP of the machine that is running the Stellarium server (in this case it is our local IP, you can use ipconfig to find it).

Example with nodejs fetch:
```js
const url = 'http://192.168.xxx.xxx:20180/api/main/view?j2000=[0.5,0.3,0.2]';
fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    }
  })
  .then(response => {
    response.text().then(text => console.log('Raw response:', text));
    return response.json();
  })
  ```

That's all you need.

Regarding the issue with "Received HTTP/0.9 when not allowed" or "invalid message type", it was resolved [here](https://github.com/Stellarium/stellarium/issues/3762).

**Reply by gzotti** [reply]
Thanks for the Node.js sample. Can I add that into the User Guide?

**Reply by FogoVoar** [reply]
> Thanks for the Node.js sample. Can I add that into the User Guide?

Sure!
